mmdet/models/dense_heads/base_dense_head.py

```diff
@@ -14,10 +18,220 @@ def loss(self, **kwargs):
         """Compute losses of the head."""
         pass
 
-    @abstractmethod
-    def get_bboxes(self, **kwargs):
-        """Transform network output for a batch into bbox predictions."""
-        pass
+    @force_fp32(apply_to=('cls_scores', 'bbox_preds'))
+    def get_bboxes(self,
+                   cls_scores,
+                   bbox_preds,
+                   score_factors=None,
+                   img_metas=None,
+                   cfg=None,
+                   rescale=False,
+                   with_nms=True,
+                   **kwargs):
+        """Transform network output for a batch into bbox predictions.
+
+        Note: score_factors is usually multiplied by the classification branch,
+        such as CenterNess in FCOS, IoU branch in ATSS.
+
+        Args:
+            cls_scores (list[Tensor]): Classification scores for all
+                scale levels, each is a 4D-tensor, the channels number
+                is num_priors * num_classes.
+            bbox_preds (list[Tensor]): Box energies / deltas for all
+                scale levels, each is a 4D-tensor, the channels number
+                is num_priors * 4.
+            score_factors (list[Tensor], Optional):  score_factor for
+                each scale level, each is a 4D-tensor, the channel number
+                is num_priors * 1. Default None.
+            img_metas (list[dict], Optional): Image meta info. Default None.
+            rescale (bool): If True, return boxes in original image space.
+                Default: False.
+            with_nms (bool): If True, do nms before return boxes.
+                Default: True.
+            cfg (mmcv.Config, Optional): Test / postprocessing configuration,
+                if None, test_cfg would be used.  Default: False.
+
+        Returns:
+            list[list[Tensor, Tensor]]: Each item in result_list is 2-tuple.
+                The first item is an (n, 5) tensor, where the first 4 columns
+                are bounding box positions (tl_x, tl_y, br_x, br_y) and the
+                5-th column is a score between 0 and 1. The second item is a
+                (n,) tensor where each item is the predicted class label of
+                the corresponding box.
+        """
+        assert len(cls_scores) == len(bbox_preds)
+
+        if score_factors is None:
+            # e.g. Retina, FreeAnchor, etc.
+            with_score_factors = False
+        else:
+            # e.g. FCOS, PAA, ATSS, etc.
+            with_score_factors = True
+            assert len(cls_scores) == len(score_factors)
+
+        num_levels = len(cls_scores)
+
+        result_list = []
+
+        for img_id in range(len(img_metas)):
+            img_meta = img_metas[img_id]
+            cls_score_list = collect_mlvl_tensor_single(cls_scores, img_id)
```

Comment:
`collect_mlvl_tensor_single` is too simple, maybe not necessary to make it as a function

Reply:
This is something worth discussing

Reply:
`collect_mlvl_tensor_single` can be renamed to `select_single_mlvl` in later PR.